blog/2023-11-14-slora.md

```diff
@@ -0,0 +1,120 @@
+
+<img src="/images/blog/slora/lora_tp.png" style="display:block; margin-top: auto; margin-left: auto; margin-right: auto; margin-bottom: auto; width: %"></img>
+<p style="color:gray; text-align: center;">Figure 3: Tensor parallelism partition strategy for batched LoRA computation.</p>
+
+The figure above demonstrates the tensor parallelism partition strategy for batched LoRA computation. This is a computational graph where nodes represent tensors/operators and the edges represent dependency. We use different colors to represent different partition strategies, which include column partition, row partition, partial sum, and replication. The per-GPU shape of each tensor is also annotated in gray. Note that $B$ is the number of tokens, $h$ is the input dimension, $N$ is the number of devices, $d$ is the hidden size, and $r$ is the adapter rank.
+
+## Methods Summary
+
+1. **Unified Paging**: To reduce memory fragmentation and increase batch size, S-LoRA introduces a unified memory pool. This pool manages dynamic adapter weights and KV cache tensors by a unified paging mechanism.
+2. **Heterogeneous Batching**: To minimize the latency overhead when batching different adapters of varying ranks, S-LoRA employs highly optimized custom CUDA kernels. These kernels operate directly on non-contiguous memory and align with the memory pool design, facilitating efficient batched inference for LoRA.
+3. **S-LoRA TP**: To ensure effective parallelization across multiple GPUs, S-LoRA introduces a novel tensor parallelism strategy. This approach incurs minimal communication cost for the added LoRA computation compared to that of the base model. This is realized by scheduling communications on small intermediate tensors and fusing the large ones with the communications of the base model.
+
+## Evaluation
+
+### Model Settings
+
+| Setting | Base model | Hidden size | Adapter ranks   |
+| ------- | ---------- | ----------- | --------------- |
+| S1      | Llama-7B   | 4096        | {8}             |
+| S2      | Llama-7B   | 4096        | {64, 32, 16, 8} |
+| S4      | Llama-13B  | 5120        | {64, 32, 16}    |
+| S5      | Llama-30B  | 7168        | {32}            |
+| S6      | Llama-70B  | 8192        | {64}            |
+
+### Baselines
+
+We compare S-LoRA with HuggingFace PEFT and vLLM.
+
+1. PEFT stands for HuggingFace PEFT: We build a server using it that batches single adapter requests and switches adapter weights between batches.
+2. vLLM-packed: Since vLLM does not support LoRA, we merge the LoRA weights into the base model and serve the multiple versions of the merged weights separately. To serve m LoRA adapters, we run `m` vLLM workers on a single GPU, where multiple workers are separate processes managed by NVIDIA MPS.
+
+### Results
```

Comment:
Reorganize this section and try to add more results.

1. Comparison with PEFT and vLLM-packed. (serving_perf.png, Table. 3)
2. Tensor parallelism (Figure. 8)
3. One subfigure from Figure 5.